lib/docsize_cache.js

```diff
@@ -0,0 +1,108 @@
+var LRU = Npm.require('lru-cache');
+var crypto = Npm.require('crypto');
+
+DocSzCache = function (maxItems, maxValues) {
+  this.items = new LRU({max: maxItems});
+  this.maxValues = maxValues;
+}
+
+DocSzCache.prototype.getSize = function (coll, query, opts, data) {
+  // If the dataset is null or empty we can't calculate the size
+  // Do not process this data and return 0 as the document size.
+  if (!(data && data.length)) {
+    return 0;
+  }
+
+  var key = this.getKey(coll, query, opts);
+  var item = this.items.get(key);
+
+  if (!item) {
+    item = new DocSzCacheItem(this.maxValues);
+    this.items.set(key, item);
+  }
+
+  if (this.needsUpdate(item)) {
+    var size = JSON.stringify(data[0]).length;
+    item.addData(size);
+  }
+
+  return item.getValue();
+};
+
+DocSzCache.prototype.getKey = function (coll, query, opts) {
+  var keyData = JSON.stringify([coll, query, opts]);
+  return crypto.createHash('md5').update(keyData).digest('hex');
+};
+
+DocSzCache.prototype.needsUpdate = function (item) {
+  // handle newly made items
+  if (!item.values.length) {
+    return true;
+  }
+
+  var currentTime = Date.now();
+  var timeSinceUpdate = currentTime - item.updated;
+  if (timeSinceUpdate > 1000*60) {
+    return true;
+  }
+
+  var cpuUsage = item.getPcpu();
+  if (cpuUsage > 80) {
```

Comment:
Even if it's greater then 80%, make sure to get some value.
But this times, it's much lower rate than earlier.